modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java

```diff
@@ -0,0 +1,237 @@
+        return delegate.booleanValueBoxed(projection[col]);
+    }
+
+    @Override
+    public byte byteValue(int col) {
+        return delegate.byteValue(projection[col]);
+    }
+
+    @Override
+    public Byte byteValueBoxed(int col) {
+        return delegate.byteValueBoxed(projection[col]);
+    }
+
+    @Override
+    public short shortValue(int col) {
+        return delegate.shortValue(projection[col]);
+    }
+
+    @Override
+    public Short shortValueBoxed(int col) {
+        return delegate.shortValueBoxed(projection[col]);
+    }
+
+    @Override
+    public int intValue(int col) {
+        return delegate.intValue(projection[col]);
+    }
+
+    @Override
+    public Integer intValueBoxed(int col) {
+        return delegate.intValueBoxed(projection[col]);
+    }
+
+    @Override
+    public long longValue(int col) {
+        return delegate.longValue(projection[col]);
+    }
+
+    @Override
+    public Long longValueBoxed(int col) {
+        return delegate.longValueBoxed(projection[col]);
+    }
+
+    @Override
+    public float floatValue(int col) {
+        return delegate.floatValue(projection[col]);
+    }
+
+    @Override
+    public Float floatValueBoxed(int col) {
+        return delegate.floatValueBoxed(projection[col]);
+    }
+
+    @Override
+    public double doubleValue(int col) {
+        return delegate.doubleValue(projection[col]);
+    }
+
+    @Override
+    public Double doubleValueBoxed(int col) {
+        return delegate.doubleValueBoxed(projection[col]);
+    }
+
+    @Override
+    public BigDecimal decimalValue(int col, int decimalScale) {
+        return delegate.decimalValue(projection[col], decimalScale);
+    }
+
+    @Override
+    public BigInteger numberValue(int col) {
+        return delegate.numberValue(projection[col]);
+    }
+
+    @Override
+    public String stringValue(int col) {
+        return delegate.stringValue(projection[col]);
+    }
+
+    @Override
+    public byte[] bytesValue(int col) {
+        return delegate.bytesValue(projection[col]);
+    }
+
+    @Override
+    public UUID uuidValue(int col) {
+        return delegate.uuidValue(projection[col]);
+    }
+
+    @Override
+    public BitSet bitmaskValue(int col) {
+        return delegate.bitmaskValue(projection[col]);
+    }
+
+    @Override
+    public LocalDate dateValue(int col) {
+        return delegate.dateValue(projection[col]);
+    }
+
+    @Override
+    public LocalTime timeValue(int col) {
+        return delegate.timeValue(projection[col]);
+    }
+
+    @Override
+    public LocalDateTime dateTimeValue(int col) {
+        return delegate.dateTimeValue(projection[col]);
+    }
+
+    @Override
+    public Instant timestampValue(int col) {
+        return delegate.timestampValue(projection[col]);
+    }
+
+    @Override
+    public ByteBuffer byteBuffer() {
+        normalizeIfNeeded();
+
+        return delegate.byteBuffer();
+    }
+
+    private void normalizeIfNeeded() {
```

Comment:
Perhaps we can go a little further with optimization.
As far as I understand, if the schema version has not been changed, we can rewrite the tuple without deserializing its contents.

```
// only in case tableRow.schemaVersion() == schemaDescriptor.version()...
private void rewriteTupleOptimized() {
	var newProjection = new int[projection.length];
	ByteBuffer tupleBuffer = delegate.byteBuffer();
	int[] requiredColumns = projection;

	var parser = new BinaryTupleParser(delegate.elementCount(), tupleBuffer);

	// Estimate total data size.
	var stats = new Sink() {
		int estimatedValueSize = 0;

		@Override
		public void nextElement(int index, int begin, int end) {
			estimatedValueSize += end - begin;
		}
	};

	for (int columnIndex : requiredColumns) {
		parser.fetch(columnIndex, stats);
	}

	// Now compose the tuple.
	BinaryTupleBuilder builder = new BinaryTupleBuilder(requiredColumns.length, stats.estimatedValueSize);

	int pos = 0;

	for (int columnIndex : requiredColumns) {
		parser.fetch(columnIndex, (index, begin, end) -> {
			if (begin == end) {
				builder.appendNull();
			} else {
				builder.appendElementBytes(tupleBuffer, begin, end - begin);
			}
		});

		newProjection[pos++] = columnIndex;
	}

	delegate = new BinaryTuple(projection.length, builder.build());
	projection = newProjection;
	normalized = true;
}
```

According to my results, this optimization only affects the performance of selectAll.

With the original changes from your PR I have the following result
```
SqlOneNodeBenchmark.selectAll    false  avgt    5  165.284 ± 13.007  ms/op
```

With this optimization
```
SqlOneNodeBenchmark.selectAll    false  avgt    5  90.457 ± 8.526  ms/op
```

Maybe it's worth thinking about adding some kind of such optimization?

Reply:
@xtern, you are correct. It definitely worth to do.

In fact, I was going to address this in the next patch, but since you've already done the job, I'm happy to include it into current patch :)